File: python/src/transtex/formatters/apa7.py
"""APA 7th formatter implementation."""
from __future__ import annotations

from typing import List

from ..reference import Reference
from .shared import author_initials, join_clauses, normalize_page_range, preferred_locator, sentence_case


def format_apa7(reference: Reference) -> str:
    """Format a reference using simplified APA 7th rules."""
    parts = [
        _author_section(reference),
        _year_section(reference),
        _title_section(reference),
        _container_section(reference),
        _locator_section(reference),
    ]
    return " ".join(part for part in parts if part).strip()


def _author_section(reference: Reference) -> str:
    raw_authors = reference.normalized_authors()
    authors = _apa7_authors(raw_authors)
    if len(raw_authors) == 1 and raw_authors[0] not in authors:
        authors = f"{authors} ({raw_authors[0]})" if authors else raw_authors[0]
    return authors


def _apa7_authors(authors: List[str]) -> str:
    formatted = author_initials(authors)
    count = len(formatted)
    if count == 0:
        return ""
    if count == 1:
        return formatted[0]
    if count <= 20:
        return ", ".join(formatted[:-1]) + f", & {formatted[-1]}"
    leading = ", ".join(formatted[:19])
    trailing = formatted[-1]
    return f"{leading}, ... {trailing}"


def _year_section(reference: Reference) -> str:
    return f"({reference.year})." if reference.year else ""


def _title_section(reference: Reference) -> str:
    title = sentence_case(reference.title or "")
    return f"{title}." if title else ""


def _container_section(reference: Reference) -> str:
    container = reference.primary_container()
    if not container:
        return ""
    volume_issue = _volume_issue(reference.volume, reference.issue)
    pages = normalize_page_range(reference.pages) or ""
    if reference.journal:
        return join_clauses([container, volume_issue, pages]) + "."
    if reference.booktitle:
        chapter_pages = f"(pp. {pages})" if pages else ""
        edition = f"({reference.edition})" if reference.edition else ""
        return join_clauses(
            [f"In {reference.booktitle}", edition, chapter_pages, reference.publisher or ""]
        ) + "."
    parts = [container]
    if reference.edition:
        parts.append(reference.edition)
    if reference.place:
        parts.append(reference.place)
    if reference.publisher and reference.publisher not in parts:
        parts.append(reference.publisher)
    if pages:
        parts.append(pages)
    if reference.accessed_date:
        parts.append(f"Retrieved {reference.accessed_date}")
    return join_clauses(parts) + "."


def _volume_issue(volume: str | None, issue: str | None) -> str:
    if volume and issue:
        return f"{volume}({issue})"
    if volume:
        return volume
    return ""


def _locator_section(reference: Reference) -> str:
    locator = preferred_locator(reference, prefix_doi="https://doi.org/")
    return locator or ""


__all__ = ["format_apa7"]

File: typescript/src/formatters/apa7.ts
import { Reference } from "../reference";
import { authorInitials, joinClauses, normalizePageRange, preferredLocator, sentenceCase } from "./shared";

export function formatApa7(reference: Reference): string {
    const parts = [
        authorSection(reference),
        yearSection(reference),
        titleSection(reference),
        containerSection(reference),
        locatorSection(reference),
    ];
    return parts.filter(Boolean).join(" ").trim();
}

function authorSection(reference: Reference): string {
    return apa7Authors(reference.normalizedAuthors());
}

function apa7Authors(authors: string[]): string {
    const formatted = authorInitials(authors);
    const count = formatted.length;
    if (count === 0) return "";
    if (count === 1) return formatted[0];
    if (count <= 20) {
        return `${formatted.slice(0, -1).join(", ")}, & ${formatted[formatted.length - 1]}`;
    }
    const leading = formatted.slice(0, 19).join(", ");
    const trailing = formatted[formatted.length - 1];
    return `${leading}, ... ${trailing}`;
}

function yearSection(reference: Reference): string {
    return reference.year ? `(${reference.year}).` : "";
}

function titleSection(reference: Reference): string {
    const title = sentenceCase(reference.title ?? "");
    if (!title) return "";
    if (reference.reportNumber) {
        return `${title} (Report No. ${reference.reportNumber}).`;
    }
    return `${title}.`;
}

function containerSection(reference: Reference): string {
    const container = reference.primaryContainer();
    if (!container) return "";
    const pages = normalizePageRange(reference.pages) ?? "";
    if (reference.journal) {
        const volumeIssueSegment = reference.volume
            ? reference.issue
                ? `${reference.volume}(${reference.issue})`
                : reference.volume
            : "";
        const body = joinClauses([container, volumeIssueSegment, pages]);
        return `${body}.`;
    }
    if (reference.booktitle) {
        const chapterPages = pages ? `(pp. ${pages})` : "";
        const edition = reference.edition ? `(${reference.edition})` : "";
        return `${joinClauses([`In ${reference.booktitle}`, edition, chapterPages, reference.publisher ?? ""])}.`;
    }
    if (reference.eventTitle) {
        const chapterPages = pages ? `(pp. ${pages})` : "";
        const location = reference.eventLocation ?? reference.place ?? "";
        return `${joinClauses([`In ${reference.eventTitle}`, chapterPages, location, reference.publisher ?? ""])}.`;
    }
    const parts = [
        container,
        reference.edition ?? "",
        reference.place ?? "",
        reference.publisher ?? "",
        pages,
        reference.accessedDate ? `Retrieved ${reference.accessedDate}` : "",
    ];
    return `${joinClauses(parts)}.`;
}

function locatorSection(reference: Reference): string {
    return preferredLocator(reference, "https://doi.org/") || "";
}

export const __all__ = ["formatApa7"];
